File: src/main/java/com/jef/justenoughfakepixel/features/misc/slotbinds/SlotBinds.java
package com.jef.justenoughfakepixel.features.misc.slotbinds;

import com.jef.justenoughfakepixel.core.JefConfig;
import com.jef.justenoughfakepixel.core.config.editors.ChromaColour;
import com.jef.justenoughfakepixel.events.SlotClickEvent;
import com.jef.justenoughfakepixel.init.RegisterEvents;
import com.jef.justenoughfakepixel.utils.chat.ChatUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.inventory.GuiInventory;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.WorldRenderer;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.inventory.Slot;
import net.minecraftforge.client.event.GuiScreenEvent;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import org.lwjgl.input.Keyboard;
import org.lwjgl.opengl.GL11;

import java.util.Iterator;
import java.util.Map;

/**
 * Slot Binds — ported from Odin's SlotBinds.kt into JEF's 1.8.9 Forge environment.
 *
 * <p>Usage:
 * <ol>
 *   <li>Open your player inventory.</li>
 *   <li>Hover a slot and press the configured <em>Bind Set Key</em> to begin a bind.</li>
 *   <li>Hover the second slot (one must be hotbar 36–44) and press the key again to save the bind.</li>
 *   <li>Hovering a bound slot and pressing the key with no pending slot removes the bind.</li>
 *   <li>Shift-click a bound slot to instantly swap it with its paired hotbar slot.</li>
 *   <li>Hold Shift while hovering a bound slot to preview the line to its pair.</li>
 * </ol>
 */
@RegisterEvents
public class SlotBinds {

    private static final Minecraft MC = Minecraft.getMinecraft();

    /** The slot index first chosen when setting a new bind; null when idle. */
    private Integer previousSlot = null;

    // ── Shift-click swap ─────────────────────────────────────────────────────

    /**
     * When the player shift-clicks a slot in their inventory that has a bind,
     * intercept and perform the hotbar swap instead of the normal shift-click.
     */
    @SubscribeEvent(priority = EventPriority.HIGH)
    public void onSlotClick(SlotClickEvent event) {
        if (!isEnabled()) return;
        if (!(event.getGui() instanceof GuiInventory)) return;
        // clickType 1 == shift-click
        if (event.getClickType() != 1) return;

        Slot slot = event.getSlot();
        if (slot == null) return;

        int slotIdx = slot.slotNumber;
        // Only intercept slots 5–44 (player inv + hotbar, excluding armour 0–4)
        if (slotIdx < 5 || slotIdx > 44) return;

        Map<Integer, Integer> binds = SlotBindStorage.getInstance().getBinds();
        Integer boundSlot = binds.get(slotIdx);
        if (boundSlot == null) return;

        // Determine which is the hotbar slot
        int from, to;
        if (slotIdx >= 36) {
            from = boundSlot;
            to = slotIdx;
        } else if (boundSlot >= 36) {
            from = slotIdx;
            to = boundSlot;
        } else {
            return; // neither is hotbar — shouldn't happen, but guard anyway
        }

        // Perform the swap via hotbar-swap click type (2) using the hotbar key number
        int hotbarKey = to - 36; // 0-8
        MC.playerController.windowClick(
                event.getGui().inventorySlots.windowId,
                from,
                hotbarKey,
                2, // SWAP
                MC.thePlayer
        );
        event.setCanceled(true);
    }

    // ── Bind set / remove ────────────────────────────────────────────────────

    @SubscribeEvent
    public void onKeyTyped(GuiScreenEvent.KeyboardInputEvent.Pre event) {
        if (!isEnabled()) return;
        if (!(event.gui instanceof GuiInventory)) return;
        if (!Keyboard.getEventKeyState()) return; // key-up, ignore

        int bindKey = JefConfig.feature.misc.slotBinds.bindKey;
        if (bindKey == Keyboard.KEY_NONE || Keyboard.getEventKey() != bindKey) return;

        Slot hovered = getHoveredSlot((net.minecraft.client.gui.inventory.GuiContainer) event.gui);
        if (hovered == null) return;

        int hoveredIdx = hovered.slotNumber;
        if (hoveredIdx < 5 || hoveredIdx > 44) return;

        event.setCanceled(true);

        Map<Integer, Integer> binds = SlotBindStorage.getInstance().getBinds();

        if (previousSlot != null) {
            int first = previousSlot;
            previousSlot = null;

            if (first == hoveredIdx) {
                ChatUtils.sendMessage("§c[SlotBinds] You can't bind a slot to itself.");
                return;
            }
            if (first < 36 && hoveredIdx < 36) {
                ChatUtils.sendMessage("§c[SlotBinds] One of the slots must be a hotbar slot (36–44).");
                return;
            }

            binds.put(first, hoveredIdx);
            binds.put(hoveredIdx, first); // bidirectional for lookup in both directions
            SlotBindStorage.getInstance().save();
            ChatUtils.sendMessage("§a[SlotBinds] Bound slot §b" + first + " §ato §d" + hoveredIdx + "§a.");
        } else {
            // If hovering a bound slot, remove it; otherwise start a new bind
            if (binds.containsKey(hoveredIdx)) {
                int partner = binds.get(hoveredIdx);
                binds.remove(hoveredIdx);
                binds.remove(partner);
                SlotBindStorage.getInstance().save();
                ChatUtils.sendMessage("§c[SlotBinds] Removed bind between slot §b" + hoveredIdx + " §cand §d" + partner + "§c.");
            } else {
                previousSlot = hoveredIdx;
                ChatUtils.sendMessage("§e[SlotBinds] Selected slot §b" + hoveredIdx + "§e. Now hover the second slot and press the key again.");
            }
        }
    }

    // ── Line preview ─────────────────────────────────────────────────────────

    /**
     * Draw a line from the hovered bound slot to its partner when the player
     * is holding Shift in the player inventory.
     * Also draws a line from {@code previousSlot} to the cursor while setting a bind.
     */
    @SubscribeEvent
    public void onDrawScreen(GuiScreenEvent.DrawScreenEvent.Post event) {
        if (!isEnabled()) return;
        if (!(event.gui instanceof GuiInventory)) return;

        GuiInventory gui = (GuiInventory) event.gui;
        Slot hovered = getHoveredSlot(gui);

        int startSlotIdx = -1;
        int endSlotIdx = -1;
        boolean useMouse = false;

        Map<Integer, Integer> binds = SlotBindStorage.getInstance().getBinds();

        if (previousSlot != null) {
            // Mid-bind: draw from first selected slot to the cursor
            startSlotIdx = previousSlot;
            useMouse = true;
        } else if (hovered != null && (Keyboard.isKeyDown(Keyboard.KEY_LSHIFT) || Keyboard.isKeyDown(Keyboard.KEY_RSHIFT))) {
            // Preview: draw from hovered bound slot to its partner
            int hovIdx = hovered.slotNumber;
            Integer partner = binds.get(hovIdx);
            if (partner == null) return;
            startSlotIdx = hovIdx;
            endSlotIdx = partner;
        } else {
            return;
        }

        // Resolve start coordinates (centre of slot)
        Slot startSlot = getSlotByIndex(gui, startSlotIdx);
        if (startSlot == null) return;

        int gx = gui.guiLeft;
        int gy = gui.guiTop;

        float x1 = gx + startSlot.xDisplayPosition + 8;
        float y1 = gy + startSlot.yDisplayPosition + 8;

        float x2, y2;
        if (useMouse) {
            x2 = event.mouseX;
            y2 = event.mouseY;
        } else {
            Slot endSlot = getSlotByIndex(gui, endSlotIdx);
            if (endSlot == null) return;
            x2 = gx + endSlot.xDisplayPosition + 8;
            y2 = gy + endSlot.yDisplayPosition + 8;
        }

        // Parse the configured colour
        String colorStr = JefConfig.feature.misc.slotBinds.lineColor;
        int argb = ChromaColour.specialToChromaRGB(colorStr);
        int a = (argb >> 24) & 0xFF;
        int r = (argb >> 16) & 0xFF;
        int g = (argb >> 8) & 0xFF;
        int b = argb & 0xFF;

        drawLine(x1, y1, x2, y2, r, g, b, a);
    }

    // ── Cleanup ───────────────────────────────────────────────────────────────

    @SubscribeEvent
    public void onGuiClose(GuiScreenEvent.InitGuiEvent event) {
        // Reset pending bind whenever a GUI is (re-)opened or closed
        previousSlot = null;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private static boolean isEnabled() {
        return JefConfig.feature != null && JefConfig.feature.misc.slotBinds.enabled;
    }

    /**
     * Retrieves the slot currently under the mouse cursor using the public
     * {@code GuiContainer#getSlotUnderMouse()} method — works in both dev and
     * production (no obfuscation issues).
     */
    private static Slot getHoveredSlot(net.minecraft.client.gui.inventory.GuiContainer gui) {
        return gui.getSlotUnderMouse();
    }

    /**
     * Finds a {@link Slot} by its {@code slotNumber} inside the GUI's container.
     */
    private static Slot getSlotByIndex(net.minecraft.client.gui.inventory.GuiContainer gui, int slotNumber) {
        for (Slot s : gui.inventorySlots.inventorySlots) {
            if (s.slotNumber == slotNumber) return s;
        }
        return null;
    }

    /**
     * Draws a 2-pixel-wide 2-D line between two screen points using the Tessellator.
     */
    private static void drawLine(float x1, float y1, float x2, float y2, int r, int g, int b, int a) {
        GlStateManager.disableTexture2D();
        GlStateManager.enableBlend();
        GlStateManager.blendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
        GlStateManager.disableLighting();
        GlStateManager.disableAlpha();
        GL11.glLineWidth(2f);

        Tessellator tess = Tessellator.getInstance();
        WorldRenderer wr = tess.getWorldRenderer();
        wr.begin(GL11.GL_LINES, DefaultVertexFormats.POSITION_COLOR);
        wr.pos(x1, y1, 0).color(r, g, b, a).endVertex();
        wr.pos(x2, y2, 0).color(r, g, b, a).endVertex();
        tess.draw();

        GL11.glLineWidth(1f);
        GlStateManager.enableAlpha();
        GlStateManager.disableBlend();
        GlStateManager.enableTexture2D();
        GlStateManager.color(1f, 1f, 1f, 1f);
    }
}

File: src/main/kotlin/io/hamlook/aetheria/features/qol/SlotBinds.kt
package io.hamlook.aetheria.features.qol

import io.hamlook.aetheria.api.event.HandleEvent
import io.hamlook.aetheria.core.ATHRConfig
import io.hamlook.aetheria.core.moulconfig.editors.ChromaColour
import io.hamlook.aetheria.events.ASMGuiDrawEvent
import io.hamlook.aetheria.events.ASMGuiInitPreEvent
import io.hamlook.aetheria.events.ASMKeyEvent
import io.hamlook.aetheria.events.ASMMouseEvent
import io.hamlook.aetheria.init.RegisterEvents
import io.hamlook.aetheria.utils.chat.ChatUtils
import io.hamlook.aetheria.utils.compat.*
import io.hamlook.aetheria.utils.data.SkyblockData
import io.hamlook.aetheria.utils.render.RenderUtils
import net.minecraft.client.gui.inventory.GuiInventory
import net.minecraft.inventory.Container
import org.lwjgl.input.Keyboard

@RegisterEvents
class SlotBinds {

    private var pendingSlot: Int? = null

    private fun cfg() = ATHRConfig.feature?.qol?.slotBinds

    private fun isEnabled(): Boolean {
        val c = cfg() ?: return false
        return c.enabled && (!c.skyblockOnly || SkyblockData.isOnSkyblock())
    }

    private fun binds() = cfg()?.binds

    private fun Int.isHotbar() = this in 36..44
    private fun Int.isArmor() = this in 5..8
    private fun Int.isValidSlot() = this in 5 until 45
    private fun isShiftDown() =
        KeyboardCompat.isKeyDown(Keyboard.KEY_LSHIFT) || KeyboardCompat.isKeyDown(Keyboard.KEY_RSHIFT)

    private fun canBind(a: Int, b: Int) = !(a.isArmor() && !b.isHotbar()) && !(b.isArmor() && !a.isHotbar())

    private fun slotCenter(container: Container, slotIndex: Int, gui: GuiInventory): Pair<Int, Int>? {
        val slot = container.getSlot(slotIndex) ?: return null
        return (slot.xDisplayPosition + gui.guiLeft + 8) to (slot.yDisplayPosition + gui.guiTop + 8)
    }

    private fun MutableMap<Int, Int>.removeBind(slot: Int) {
        val partner = remove(slot)
        if (partner != null) remove(partner)
    }

    private fun MutableMap<Int, Int>.addBind(a: Int, b: Int) {
        removeBind(a)
        removeBind(b)

        this[a] = b
        this[b] = a
    }

    @HandleEvent(priority = HandleEvent.HIGHEST)
    fun onMouseClick(event: ASMMouseEvent) {
        if (!isEnabled() || !isShiftDown()) return
        val gui = event.gui as? GuiInventory ?: return
        if (!MouseCompat.getEventButtonState() || MouseCompat.getEventButton() != 0) return
        val clicked = InventoryCompat.getSlotUnderMouse(gui)?.slotNumber?.takeIf { it.isValidSlot() } ?: return
        val bound = binds()?.get(clicked) ?: return
        if (!bound.isValidSlot() || bound == clicked) return

        val (from, to) = if (clicked.isHotbar()) bound to clicked else if (bound.isHotbar()) clicked to bound else return

        InventoryCompat.windowClick(
            InventoryCompat.getContainer(gui).windowId, from, to - 36, 2, MinecraftCompat.getLocalPlayer() ?: return
        )
        event.cancel()
    }

    @HandleEvent
    fun onKeyPress(event: ASMKeyEvent) {
        if (!isEnabled()) return
        val gui = event.gui as? GuiInventory ?: return
        val c = cfg() ?: return
        if (c.bindKey == Keyboard.KEY_NONE || KeyboardCompat.getEventKey() != c.bindKey || !KeyboardCompat.getEventKeyState()) return

        val clicked = InventoryCompat.getSlotUnderMouse(gui)?.slotNumber?.takeIf { it.isValidSlot() } ?: return
        event.cancel()

        val pending = pendingSlot
        if (pending != null) {
            pendingSlot = null
            when {
                pending == clicked -> ChatUtils.sendMessage("§cCan't bind a slot to itself.")
                !pending.isHotbar() && !clicked.isHotbar() -> ChatUtils.sendMessage("§cOne slot must be in the hotbar.")
                !canBind(pending, clicked) -> ChatUtils.sendMessage("§cArmor slots can only be bound to hotbar slots.")
                else -> {
                    c.binds.addBind(pending, clicked)
                    ATHRConfig.saveConfig()
                }
            }
        } else {
            if (c.binds.containsKey(clicked)) {
                c.binds.removeBind(clicked)
                ATHRConfig.saveConfig()
            } else {
                pendingSlot = clicked
            }
        }
    }

    @HandleEvent
    fun onDraw(event: ASMGuiDrawEvent) {
        if (!isEnabled()) return
        val gui = event.gui as? GuiInventory ?: return
        val c = cfg() ?: return
        val container = InventoryCompat.getContainer(gui) ?: return
        val color = ChromaColour.specialToChromaRGB(c.lineColor)

        GlStateManagerCompat.pushMatrix()
        GlStateManagerCompat.translate(0.0, 0.0, 999.0)

        try {
            val pending = pendingSlot
            if (pending != null) {
                val (sx, sy) = slotCenter(container, pending, gui) ?: return
                RenderUtils.drawLine(sx, sy, event.mouseX, event.mouseY, color, 2f)
                return
            }

            if (c.alwaysShowLines) {
                // Draw all binds, iterate unique pairs (avoid drawing Aâ†’B and Bâ†’A twice)
                val drawn = mutableSetOf<Int>()
                for ((from, to) in c.binds) {
                    if (drawn.contains(from)) continue
                    drawn.add(from)
                    drawn.add(to)
                    val (sx, sy) = slotCenter(container, from, gui) ?: continue
                    val (ex, ey) = slotCenter(container, to, gui) ?: continue
                    RenderUtils.drawLine(sx, sy, ex, ey, color, 2f)
                }
            } else {
                if (!isShiftDown()) return
                val hovered = InventoryCompat.getSlotUnderMouse(gui)?.slotNumber?.takeIf { it.isValidSlot() } ?: return
                val bound = c.binds[hovered]?.takeIf { it.isValidSlot() && it != hovered } ?: return
                val (sx, sy) = slotCenter(container, hovered, gui) ?: return
                val (ex, ey) = slotCenter(container, bound, gui) ?: return
                RenderUtils.drawLine(sx, sy, ex, ey, color, 2f)
            }
        } finally {
            GlStateManagerCompat.popMatrix()
        }
    }

    @HandleEvent
    fun onGuiClose(event: ASMGuiInitPreEvent) {
        if (event.gui == null) pendingSlot = null
    }
}
